build-tools/packages/build-cli/test/commands/typetests.test.ts

```diff
@@ -2,11 +2,330 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-import { assert } from "chai";
+import { assert, expect } from "chai";
+import clonedeep from "lodash.clonedeep";
 
-import { previousVersion } from "../../src/commands/typetests";
+import {
+	TypeTestConfigActions,
+	VersionOptions,
+	previousVersion,
+	updateTypeTestConfiguration,
+} from "../../src/commands/typetests";
+import { PackageJson } from "@fluidframework/build-tools";
+
+/**
+ * A minimal test package.json. It defines only the required fields according to the type definition.
+ */
+const packageMinimal: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+};
+
+/**
+ * A test package.json with an enabled typeValidation node.
+ */
+const packageWithTypeValidation: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+	devDependencies: {
+		"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+	},
+	typeValidation: {
+		broken: {
+			"broken-package": {
+				backCompat: false,
+				forwardCompat: false,
+			},
+		},
+		disabled: false,
+	},
+};
+
+/**
+ * A test package.json with an disabled typeValidation node.
+ */
+const packageWithTypeValidationDisabled: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+	typeValidation: {
+		broken: {
+			"broken-package": {
+				backCompat: false,
+				forwardCompat: false,
+			},
+		},
+		disabled: true,
+	},
+};
+
+const optionsMatrix: Map<string, TypeTestConfigActions> = new Map([
+	[
+		"previousAndResetTrue",
+		{
+			version: VersionOptions.Previous,
+			resetBroken: true,
+		},
+	],
+	[
+		"previousAndResetFalse",
+		{
+			version: VersionOptions.Previous,
+			resetBroken: false,
+		},
+	],
+	[
+		"clearAndResetTrue",
+		{
+			version: VersionOptions.Clear,
+			resetBroken: true,
+		},
+	],
+	[
+		"clearAndResetFalse",
+		{
+			version: VersionOptions.Clear,
+			resetBroken: false,
+		},
+	],
+	[
+		"clearIfDisabledAndResetTrue",
+		{
+			version: VersionOptions.ClearIfDisabled,
+			resetBroken: true,
+		},
+	],
+	[
+		"clearIfDisabledAndResetFalse",
+		{
+			version: VersionOptions.ClearIfDisabled,
+			resetBroken: false,
+		},
+	],
+]);
 
 describe("typetests tests", () => {
+	describe("updateTypeTestConfiguration", () => {
+		const testPackages = [
+			packageMinimal,
+			packageWithTypeValidation,
+			packageWithTypeValidationDisabled,
+		];
+
+		describe("previousAndResetTrue", () => {
+			const name = "previousAndResetTrue";
+			const options = optionsMatrix.get(name)!;
+			it(`packageMinimal`, () => {
+				const pkg = clonedeep(packageMinimal);
+				updateTypeTestConfiguration(pkg, options);
+				expect(pkg.typeValidation).to.be.undefined;
+				expect(pkg.typeValidation?.broken).to.be.undefined;
+				expect(pkg.typeValidation?.disabled).to.be.undefined;
+				expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
```

Comment:
no reason to check something exists before checking it has a specifc value. the quality check will handle undefiend just fine.

Reply:
Won't there be better error messages though?

Reply:
I think using deepEqual will address my "better error messages" problem and is cleaner overall.